src/theory/strings/extf_solver.cpp

```diff
@@ -93,77 +94,72 @@ bool ExtfSolver::doReduction(int effort, Node n)
     Trace("strings-extf-debug") << "...skip due to reduced" << std::endl;
     return false;
   }
-  // determine the effort level to process the extf at
-  // 0 - at assertion time, 1+ - after no other reduction is applicable
-  int r_effort = -1;
   // polarity : 1 true, -1 false, 0 neither
   int pol = 0;
   Kind k = n.getKind();
   if (n.getType().isBoolean() && !d_extfInfoTmp[n].d_const.isNull())
   {
     pol = d_extfInfoTmp[n].d_const.getConst<bool>() ? 1 : -1;
   }
-  if (k == STRING_CONTAINS)
+  // determine if it is the right effort
+  if (k == STRING_SUBSTR || (k == STRING_CONTAINS && pol == 1))
   {
-    if (pol == 1)
+    // we reduce these semi-eagerly, at effort 1
+    if (effort != 1)
     {
-      r_effort = 1;
+      return false;
     }
-    else if (pol == -1)
+  }
+  else if (k == STRING_CONTAINS && pol == -1)
+  {
+    // negative contains reduces at level 2, or 3 if guessing model
+    int reffort = options().strings.stringModelBasedReduction ? 3 : 2;
+    if (effort == reffort)
     {
-      if (effort == 2)
+      Node x = n[0];
+      Node s = n[1];
+      std::vector<Node> lexp;
+      Node lenx = d_state.getLength(x, lexp);
+      Node lens = d_state.getLength(s, lexp);
+      if (d_state.areEqual(lenx, lens))
       {
-        Node x = n[0];
-        Node s = n[1];
-        std::vector<Node> lexp;
-        Node lenx = d_state.getLength(x, lexp);
-        Node lens = d_state.getLength(s, lexp);
-        if (d_state.areEqual(lenx, lens))
-        {
-          Trace("strings-extf-debug")
-              << "  resolve extf : " << n
-              << " based on equal lengths disequality." << std::endl;
-          // We can reduce negative contains to a disequality when lengths are
-          // equal. In other words, len( x ) = len( s ) implies
-          //   ~contains( x, s ) reduces to x != s.
-          if (!d_state.areDisequal(x, s))
-          {
-            // len( x ) = len( s ) ^ ~contains( x, s ) => x != s
-            lexp.push_back(lenx.eqNode(lens));
-            lexp.push_back(n.negate());
-            Node xneqs = x.eqNode(s).negate();
-            d_im.sendInference(
-                lexp, xneqs, InferenceId::STRINGS_CTN_NEG_EQUAL, false, true);
-          }
-          // this depends on the current assertions, so this
-          // inference is context-dependent
-          d_extt.markReduced(n, ExtReducedId::STRINGS_NEG_CTN_DEQ, true);
-          return true;
-        }
-        else
+        Trace("strings-extf-debug")
+            << "  resolve extf : " << n
+            << " based on equal lengths disequality." << std::endl;
+        // We can reduce negative contains to a disequality when lengths are
+        // equal. In other words, len( x ) = len( s ) implies
+        //   ~contains( x, s ) reduces to x != s.
+        if (!d_state.areDisequal(x, s))
```

Comment:
We could consider making this inference regardless of `stringModelBasedReduction`, I think.

Reply:
True, but I don't think its necessary.

Actually, I am considering making disequality handling also deferred to model effort.

Notice that `x != s` would only be useful if `x = s`, but in that case we should get `contains(x,s) ---> true`.